feat(gatsby): Warn when pageContext is over 500k

evading1998 opened the pull request:
<!--
  Have any questions? Check out the contributing docs at https://gatsby.dev/contribute, or
  ask in this Pull Request and a Gatsby maintainer will be happy to help :)
-->

<!--
  Is this a blog post? Check out the docs at https://www.gatsbyjs.org/contributing/blog-contributions/, and please mention if the blog post is pre-approved
  by someone from Gatsby.
-->

## Description

updated public.js to send a warning when pageContext exceeds approx 500kb limit

### Documentation

<!--
  Where is this feature or API documented?

  - If docs exist:
    - Update any references, if relevant. This includes Guides and Gatsby Internals docs.
  - If no docs exist:
    - Create a stub for documentation including bullet points for how to use the feature, code snippets (including from happy path tests), etc.
  - Tag @gatsbyjs/learning for review, pairing, polishing of the documentation
-->

## Related Issues


  Fixes #14213

marcysutton commented:
Cool!! Are there any tests that could be added to this PR?

pieh commented:
Re the performance:

The `createPage` action creator can be called multiple times for the same page (we have `onCreatePage` hook that let users and plugins override various page properties - including `context`). So current implementation can potentially run `JSON.stringify` on same context multiple times. I don't think we should track context objects to know if we run it already (as user can just mutate it, and then tracking just won't work) + we already see scaling issues with approaches like that (node tracking WeakMap).

What at least we should try to do is let the `onCreatePage` API runs cascade to finish before trying to check to size of context?

We also already somewhat run `JSON.stringify` on `context` - just not directly - we do this whenever we run page queries:
https://github.com/gatsbyjs/gatsby/blob/f2c1be03a2a52e4f79315e35b9535637fcf8dc8a/packages/gatsby/src/query/query-runner.js#L128
(`result` will have `pageContext` field, but it will also have actual `data` as well, so there would be at least some trickiness there)

Question is - can we in some way reuse fact that we already `JSON.stringify` context in different place? Can we do this check in `query-runner` or it's just too late and the warning is just so far of the initial action that should trigger warning that it would be just confusing?

pvdz commented:
I would not like to do this by default. This doesn't scale well.

What you could do is add this check when persisting the redux cache in gatsby/packages/gatsby/src/redux/persist.ts . In the `prepareCacheFolder` function the nodes are serialized in groups. You could add a heuristic for ratio and issue a warning if the chunksize is below a certain number.

Also, using `JSON.stringify` on nodes is not "safe" because of circular references. This is why we use `v8.serialize()`, which has other drawbacks.

Sorry, this is not a "free" operation, certainly not on large objects/nodes and certainly not at scale (consider doing this a million times, multiple times).

Fwiw, I would like this kind of check for a "triage" mode, which would run slower but emit more hints about potential problems with a build. Under such a mode I have no problems with this, provided the crash potential is caught (`try/catch`). Alas, we don't have a "gatsby build --triage" yet. That only exists in my head right now.

vladar commented:
@pieh 

> Question is - can we in some way reuse fact that we already JSON.stringify context in a different place?

Maybe we should warn when both context AND data are huge? So the whole `result` and `page-data.json`? It serves a bit different purpose though than warn on just the context.

pieh commented:
> Maybe we should warn when both context AND data are huge? So the whole `result` and `page-data.json`? It serves a bit different purpose though than warn on just the context.

Yeah, it's different - this part is somewhat captured in https://github.com/gatsbyjs/gatsby/issues/19780

The context alone however have some special properties that are not shared with `page-data.json` files:

- `context` is always kept in memory (as it's part of page objects stored in redux store)
- `context` is getting inferred right now (with large amount of pages and/or large context objects - this can get quite slow)

evading1998 commented:
Thanks for the suggestions. We are working on it now

evading1998 commented:
One of the issues noted was that because createPage action is being called multiple times and we should try to avoid calling the warning before the page is created. So if we’re correct, @pvdz  suggests to instead have the warning be in persist.ts file, so that instead of creating the warning on page creation, to rather do it when caching the page in the persist storage. Is this what we should follow. If this is the case, we have a question about one of the suggestions made:
> What you could do is add this check when persisting the redux cache in gatsby/packages/gatsby/src/redux/persist.ts . In the `prepareCacheFolder` function the nodes are serialized in groups. You could add a heuristic for ratio and issue a warning if the chunksize is below a certain number.
>
Are we checking if the chunksize is below or above a certain number? From my understanding, the chunksize is directly proportional to the pageContext, so if we are checking if the page context is too big, we want to see if the corresponding chunksize is too big? 
Also, about the verbose option, how could we read in the user's choice?

KyleAMathews commented:
Perhaps we do this post build by scanning the public data folder for large files? That'd be cheaper.

pvdz commented:
@evading1998 your original PR tried to warn ahead of time, that's nice but I don't think we'd be shipping by default because of the performance implication. So the best next thing is to do it after the fact and make use of artifacts already generated anyways.

One of those artifacts is the heuristic for persisting nodes in the cache. We don't check every node but rather pick a bunch of them and hope that they represent the set (they may completely miss the mark). So the chunksize is an indicator and would work if one of these checked nodes are actually the heavy kind. But if you had a million nodes and only two bad apples (with large contexts), this heuristic would probably miss. ymmv, but the check is cheap so that's fine.

Kyle's suggestion is to do an `ls -al` on the output folders (public/.cache) and report generated pages that are large. This might work too but the drawback is that there is no guarantee since you could put a ton of data in the context without actually putting that in a page. And pages might be large while nodes have a small context (like files that paginate a bunch of other pages).

Then there's the posisbility for a "triage" mode, where we care less about perf and where we would start triaging for common problems like this. This would be a separate project. That would be a great addition and if you're interested in working on that sort of project I can help you get started on the kinds of problems such a mode should try to find.

For this PR, I think it would be good to 

- warn if the `chunkSize` drops below a certain number. You'll have to experiment a bit what that number would be. 
- doing a post-build scan through the output for large files and warning for that would also be a good thing. 

I would suggest to have these changes in separate commits (one can be in this PR the other in a new one).

What do you think?

evading1998 commented:
We think doing a post-build scan would be the most effective option here, too. We are capable of building a parser and check the page-data to see if any of them exceed a certain limit, but we are not entirely sure where we should build this, though. Is there a runner file that goes through the pageData already? Or should we create a new one, and where should the file be located?

pvdz commented:
Hi @evading1998 :) By "building a parser", do you mean something to process the `page-data.json`? Becayse I suspect that doesn't require a whole parser, but rather a quick scan. I'd rather not have anything heavier in our main build pipeline so this is why I'm asking.

Once you have any suspicious large `page-data.json` files through `find` (or whatever), you can either immediately report that some are big, or do go the extra mile to verify that the context is big, like `data = fs.readFileSync(pagedata, 'utf8'); if (data.length - data.indexOf('pageContext') > WARN_SIZE) reporter.warn('Large file found')`. Or something in that vain, anyways.

I'm sorry if that's what you already had in mind. I just don't want you to waste a lot of time on this :)

evading1998 commented:
Hello. Yes, we think doing a scan is a good idea. But we are wondering what specific file should this scan live in, since it essentially needs to check all page-data.json file. We could not decide where to put it. @pvdz

pvdz commented:
I believe a call is being setup so let's use that opportunity to discuss this a bit :)

pvdz commented:
After the call we agreed adding a warning in persist.ts would be good low hanging fruit. The message should be emit once per run.

While we did discuss checking for large files in the generated folders, it turns out that this is not feasible because there's no single file that contains all the data that goes into a page and just scannign for the size of the built pages/bundles is too conflated. So that plugin idea is moot for now.

evading1998 commented:
Hello, we just commited our changes to the persist.ts file and removed our previous changes to public.js file. We added a function that sends a warning once if any of the chunkSizes exceeds approx 500kb limit. We are not sure if this requires a jest. If that's the case, we are not very familiar with creating a jest. Maybe someone else could take it from here?

pvdz commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -81,10 +81,20 @@ function guessSafeChunkSize(values: [string, IGatsbyNode][]): number {
+
   for (let i = 0; i < valueCount; i += step) {
     const size = v8.serialize(values[i]).length
+    if (size > 500000) {
```
The `maxSize` already tracks the size of the largest node, so why would you check this inside the loop, and not check `maxSize` instead after the loop?

evading1998 commented:
Hello, we just commited our changes to the persist.ts file. We deleted the redundant loop and ran prettier on the file. The format should be correct.

pvdz commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -81,10 +81,16 @@ function guessSafeChunkSize(values: [string, IGatsbyNode][]): number {
   const valueCount = values.length
   const step = Math.max(1, Math.ceil(valueCount / nodesToTest))
   let maxSize = 0
+  
```
It's a minor thing, but could you please drop this new line as well?

Generally you'd want to create as small a diff as possible. In this case this empty line is not related to this PR so it's best to drop those kinds of changes before submitting a patch / PR.
- The fewer changes in a PR the easier it is for a reviewer to review (less cognitive overhead)
- Unnecessary changes affect the git history, which in turn unnecessarily affects bisecting problems and `git blame`

In that respect, the new empty line 80 is fine since it's part of your patch.

pvdz commented:
Oh and also, you did not run linting (`yarn lint`) because you didn't run the autoformatter Prettier (`yarn format`).

I can tell because the tests are failing for the reason of linting and that's because you left a semi-colon. In this project, the style guide tries to use as few semi-colons as possible. Regardless of the discussion around that, this is something that's required and enforced through tooling.

I'd encourage you to click through on the reported failures and see the reports. Again, you can fix this by running Prettier (for example, by running `yarn format`).

evading1998 commented:
@pvdz Thank you for the suggestions. We made the updates accordingly.

AishaBlake commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -86,6 +86,13 @@ function guessSafeChunkSize(values: [string, IGatsbyNode][]): number {
+  // Sends a warning once if any of the chunkSizes exceeds approx 500kb limit
+  if (maxSize > 500000) {
+    report.warn(
+      `The size of least one page context exceeded 500kb, this could lead to degraded performance. Consider putting less data in the page context.`
```
```suggestion
      `The size of at least one page context chunk exceeded 500kb, which could lead to degraded performance. Consider putting less data in the page context.`
```

This could use a couple of tweaks to the grammar but I think it’s a great addition! As a suggestion, I’ve also added the word “chunk” in an attempt to clarify what’s being measured.

evading1998 commented:
Thank you for the suggestions. We had updated the file accordingly. @AishaBlake @pvdz

Blumed commented:
This is great! Very handy. My one gripe is the message. It doesn't tell me which file or how many. It is a bit vague. At the very least it would be nice to communicate where to look for the page-data.json's flagged. Like a file path do the dir's.

Keep up the awesome work team!

phun-ky commented:
Yeah, I am in this situation now @Blumed , I get this warning, but I can't track down the culprit for this message..
